File: identity/go/kernaq/verifications.go
package kernaq

import (
	"context"
	"fmt"
	"io"
)

// VerifyResource handles the synchronous KYC pipeline.
// POST /v1/verify and POST /v1/verify/sandbox.
type VerifyResource struct {
	c *client
}

// VerifyInput is the input to Verify.Run and Verify.Sandbox.
type VerifyInput struct {
	// Document is the government-issued ID photo (JPEG, PNG, PDF).
	Document         io.Reader
	DocumentFilename string // default: "document.jpg"

	// Selfie is the portrait face photo.
	Selfie         io.Reader
	SelfieFilename string // default: "selfie.jpg"

	// Liveness: supply Video OR all three Frame fields (low-bandwidth).
	Video         io.Reader
	VideoFilename string // default: "liveness.mp4"

	Frame1         io.Reader
	Frame1Filename string
	Frame2         io.Reader
	Frame2Filename string
	Frame3         io.Reader
	Frame3Filename string

	// DocumentType: "national_id", "passport", "driver_license", etc.
	DocumentType string
	// Country: ISO 3166-1 alpha-3, e.g. "KEN".
	Country string
}

// VerifyResult is returned synchronously by Run and Sandbox.
// Built in-memory by the server, returned in the HTTP response, never stored.
type VerifyResult struct {
	RequestID      string         `json:"request_id"`
	Verdict        string         `json:"verdict"`          // "pass" | "fail" | "review"
	Score          int            `json:"score"`            // 0–100
	FaceMatch      bool           `json:"face_match"`
	FaceConfidence float64        `json:"face_confidence"`
	LivenessPass   bool           `json:"liveness_pass"`
	DocumentFields DocumentFields `json:"document_fields"`
	FraudFlags     []string       `json:"fraud_flags"`
	FailureReason  string         `json:"failure_reason,omitempty"`
	DurationMs     int64          `json:"duration_ms"`
}

// Run submits a full KYC verification in live mode.
// Deducts one credit. Blocks 3–8 seconds. Returns the result directly.
// Nothing is stored on Kernaq servers.
//
// Example:
//
//	result, err := client.Verify.Run(ctx, kernaq.VerifyInput{
//	    Document:     docFile,
//	    Selfie:       selfieFile,
//	    Video:        videoFile,
//	    DocumentType: "national_id",
//	    Country:      "KEN",
//	})
//	if err != nil { ... }
//	if result.Verdict == "pass" {
//	    fmt.Println(result.DocumentFields.Name)
//	}
func (r *VerifyResource) Run(ctx context.Context, input VerifyInput) (*VerifyResult, error) {
	return r.submit(ctx, "/verify", input)
}

// Sandbox submits a full KYC verification in sandbox mode.
// Same pipeline, no billing. Use k_test_ API keys.
func (r *VerifyResource) Sandbox(ctx context.Context, input VerifyInput) (*VerifyResult, error) {
	return r.submit(ctx, "/verify/sandbox", input)
}

func (r *VerifyResource) submit(ctx context.Context, path string, input VerifyInput) (*VerifyResult, error) {
	fields := map[string]string{
		"document_type": input.DocumentType,
		"country":       input.Country,
	}

	docFilename  := nameOr(input.DocumentFilename, "document.jpg")
	selfFilename := nameOr(input.SelfieFilename,   "selfie.jpg")

	files := []filePart{
		{field: "document", reader: input.Document, filename: docFilename,  mime: mimeFromName(docFilename)},
		{field: "selfie",   reader: input.Selfie,   filename: selfFilename, mime: mimeFromName(selfFilename)},
	}

	if input.Video != nil {
		vidFilename := nameOr(input.VideoFilename, "liveness.mp4")
		files = append(files, filePart{
			field: "video", reader: input.Video, filename: vidFilename, mime: mimeFromName(vidFilename),
		})
	} else if input.Frame1 != nil && input.Frame2 != nil && input.Frame3 != nil {
		f1 := nameOr(input.Frame1Filename, "frame1.jpg")
		f2 := nameOr(input.Frame2Filename, "frame2.jpg")
		f3 := nameOr(input.Frame3Filename, "frame3.jpg")
		files = append(files,
			filePart{field: "frame1", reader: input.Frame1, filename: f1, mime: "image/jpeg"},
			filePart{field: "frame2", reader: input.Frame2, filename: f2, mime: "image/jpeg"},
			filePart{field: "frame3", reader: input.Frame3, filename: f3, mime: "image/jpeg"},
		)
	}

	var out VerifyResult
	if err := r.c.upload(ctx, path, fields, files, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// UsageResource handles GET /v1/usage and GET /v1/logs.
type UsageResource struct {
	c *client
}

// Get returns aggregate non-PII usage statistics.
// days: number of days to include (1–365, default 30).
func (r *UsageResource) Get(ctx context.Context, days int) (*UsageSummary, error) {
	if days <= 0 {
		days = 30
	}
	var out UsageSummary
	if err := r.c.get(ctx, fmt.Sprintf("/usage?period=%dd", days), &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// GetLogs returns paginated raw API call logs (no PII).
// Each entry has endpoint, status_code, duration_ms, created_at.
func (r *UsageResource) GetLogs(ctx context.Context, limit, offset int) (*LogsResponse, error) {
	if limit <= 0 {
		limit = 50
	}
	var out LogsResponse
	if err := r.c.get(ctx, fmt.Sprintf("/logs?limit=%d&offset=%d", limit, offset), &out); err != nil {
		return nil, err
	}
	return &out, nil
}

File: identity/node/src/verifications.ts
import { BaseClient, type FileInput } from './client.js'
import type { VerifyRequest, VerifyResult } from './types.js'

export class VerificationsResource extends BaseClient {

  /**
   * Run a full live KYC verification. Charges a credit.
   * Blocks 3–8 seconds, returns the result directly.
   *
   * Liveness: supply `video` OR `frame1` + `frame2` + `frame3`.
   *
   * @example
   * const result = await kernaq.verify.run({
   *   document:     fs.createReadStream('id.jpg'),
   *   selfie:       fs.createReadStream('selfie.jpg'),
   *   video:        fs.createReadStream('liveness.mp4'),
   *   documentType: 'national_id',
   *   country:      'KEN',
   * })
   * if (result.verdict === 'pass') { ... }
   */
  run(req: VerifyRequest): Promise<VerifyResult> {
    return this.uploadVerify('/verify', req)
  }

  /**
   * Run verification through the sandbox pipeline — same logic, no billing.
   *
   * @example
   * const result = await kernaq.verify.sandbox({
   *   document:     fs.createReadStream('id.jpg'),
   *   selfie:       fs.createReadStream('selfie.jpg'),
   *   video:        fs.createReadStream('liveness.mp4'),
   *   documentType: 'passport',
   *   country:      'GBR',
   * })
   */
  sandbox(req: VerifyRequest): Promise<VerifyResult> {
    return this.uploadVerify('/verify/sandbox', req)
  }

  private uploadVerify(path: string, req: VerifyRequest): Promise<VerifyResult> {
    const fields: Record<string, string> = {
      document_type: req.documentType,
      country:       req.country,
    }

    const files: Array<{ field: string; value: FileInput; filename: string; contentType: string }> = [
      {
        field:       'document',
        value:       req.document as FileInput,
        filename:    req.documentName ?? 'document.jpg',
        contentType: contentTypeFromName(req.documentName ?? 'document.jpg'),
      },
      {
        field:       'selfie',
        value:       req.selfie as FileInput,
        filename:    req.selfieName ?? 'selfie.jpg',
        contentType: contentTypeFromName(req.selfieName ?? 'selfie.jpg'),
      },
    ]

    if (req.video) {
      files.push({
        field:       'video',
        value:       req.video as FileInput,
        filename:    req.videoName ?? 'liveness.mp4',
        contentType: contentTypeFromName(req.videoName ?? 'liveness.mp4'),
      })
    } else if (req.frame1 && req.frame2 && req.frame3) {
      files.push(
        { field: 'frame_1', value: req.frame1 as FileInput, filename: req.frame1Name ?? 'frame_1.jpg', contentType: 'image/jpeg' },
        { field: 'frame_2', value: req.frame2 as FileInput, filename: req.frame2Name ?? 'frame_2.jpg', contentType: 'image/jpeg' },
        { field: 'frame_3', value: req.frame3 as FileInput, filename: req.frame3Name ?? 'frame_3.jpg', contentType: 'image/jpeg' },
      )
    }

    return this.upload<VerifyResult>(path, fields, files)
  }
}

function contentTypeFromName(filename: string): string {
  const ext = filename.split('.').pop()?.toLowerCase() ?? ''
  const map: Record<string, string> = {
    jpg: 'image/jpeg', jpeg: 'image/jpeg',
    png: 'image/png',  heic: 'image/heic',
    pdf: 'application/pdf',
    mp4: 'video/mp4',  mov: 'video/quicktime', webm: 'video/webm',
  }
  return map[ext] ?? 'application/octet-stream'
}
